### ISSUE TITLE
(shared) Memory leak on Pytorch 1.0

### ISSUE THREAD

**Issue by snippler** [pytorch/pytorch#17499]
## 🐛 Bug

I’d like to report a weird behaviour in 1.0 that I could resolve by only going back to 0.4. I am training a network on quite a big data set (35 GB) and use 4 gpus by applying the command
torch.nn.DataParallel(model).cuda()
Further I am having a big batch size (>1000) which makes the command
torch.multiprocessing.set_sharing_strategy(‘file_system’). I have num_workers=16 in the dataloader
necessary. Now the trouble begins: every epoch my /dev/shm increases by ca. 3GB. At some point it is full and my process crashes. I tried 1.0.0 and 1.0.1 but both showed this bahaviour. Pytorch 0.4 does not have this problem, /dev/shm is never above 1GB.

## To Reproduce

1. Train network on big data set with data.Dataloader with big batch size, for which you require torch.multiprocessing.set_sharing_strategy('file_system') and Dataparallel
2. Observe /dev/shm until it is full

## Expected behavior

In 0.4 version /dev/shm is not filling up and there is not crash

 - PyTorch Version (e.g., 1.0.1 and 1.0.0 vs. 0.4.0):
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source):
 - Python version: 2.7
 - CUDA/cuDNN version: 10
 - GPU models and configuration:
 - Any other relevant information:

**Comment by otaj** [comment]
Seems it might be related to this issue #17466 
AFAIK, `__hash__` is what changed in between 1.0.0 and 0.4.0

**Comment by snippler** [comment]
I tested with shuffle=False and /dev/shm is still accumulating so it is not related.

**Comment by otaj** [comment]
I meant related as in DataParallel or set_sharing_strategy('file_system') are using tensors as keys in some dict, thinking that `__hash__` is value-dependent while it actually is only memory-location-dependent. Which of course is a bug on its own, I'm just trying to help pinpoint the origin of the bug

**Comment by snippler** [comment]
I see your point.
Any idea how to test this?

**Comment by otaj** [comment]
Honestly, no idea. Try to run this whole thing inside some debugger, and look inside attributes of DataParallel, that's how I tracked down my problem. The bug should manifest regardless of the dataset size, therefore debug it on some small subset, which should definitely speed up the whole debugging process

**Comment by snippler** [comment]
In the code that you provided you found out that key is a torch.tensor?

I am not able to reproduce this. Printing key I get an integer with 0.4 and with 1.0. Do I misunderstand something?

**Comment by otaj** [comment]
Yes, the key passed to the `__getitem__` of my dataset by DataLoader with shuffle=True (therefore essentially using RandomSampler) was an instance of torch.tensor (0-D, LongTensor, however torch.tensor).

**Comment by otaj** [comment]
However, it seems, that the main issue is from random_split. To reproduce:

  class datalist(list):
      def __getitem__(self, key):
          print(type(key))
          return super().__getitem__(key)

  import torch.utils.data as data
  dataset = datalist(range(10))
  n_dataset = data.random_split(dataset, [10])
  for batch in data.DataLoader(n_dataset):
      pass # <- this will print torch.tensor, cuplrit seems to be random_split ten times because of monkey-patched __getitem__

**Comment by pietern** [comment]
Per the report in #24200 marking this issue high priority as well.

**Comment by Baranowski** [comment]
@zou3519, are you working on this or trying to repro?

**Comment by zou3519** [comment]
I'm not actively working on it so I am unassigning myself for now.

**Comment by Baranowski** [comment]
(note to self) Possibly related: https://github.com/pytorch/pytorch/issues/27218

**Comment by Baranowski** [comment]
@snippler, I know it's been over a year but do you have a code that reproduces this issue?

I have looked at the DataParallel and DataLoader code and don't see any obvious place where they would trigger the __hash__ issue.

I've also ran https://github.com/mrcolo/research_srgan which uses both DataParallel and DataLoader, and I didn't see increase in /dev/shm usage between epochs.

**Comment by snippler** [comment]
I am not sure if I still have the code where the problem could be reproduced. Probably not but I'll post if I find it.

**Comment by Baranowski** [comment]
Closing since I cannot reproduce, and it might be a duplicate of one of the issues linked above.
